### ISSUE TITLE
More insight on how :local and :global works in css modules

### ISSUE THREAD

**Issue by jgfresh** [css-modules/css-modules#264]
Hi, I'm using css modules however I still don't know how to use :local(.class) and :global(.class)

All of the classes are local by default so what are we exactly saying by doing :local(.class) ?

**Comment by dfdeagle47** [comment]
My understanding is that it does nothing if the classes are local by default. It's just used if you have the global scope as the default.

https://github.com/webpack-contrib/css-loader#scope

> By default CSS exports all classnames into a global selector scope. Styles can be locally scoped to avoid globally scoping styles.
>
> The syntax :local(.className) can be used to declare className in the local scope. The local identifiers are exported by the module.
>
> With :local (without brackets) local mode can be switched on for this selector. :global(.className) can be used to declare an explicit global selector. With :global (without brackets) global mode can be switched on for this selector.

**Comment by douglasg14b** [comment]
Hm, that doesn't clarify it too much, and damn this thread was hard to find. Examples of what the effects of :local and :global are?

**Comment by douglasg14b** [comment]
Found this: https://github.com/webpack-contrib/css-loader#scope

**Comment by liuliangsir** [comment]
That's absolutely right, all classes are local by default. What if you would like to apply global scope to this selector inside a local selector? In this case,  you can use the syntax `:global (without brackets)` for this selector above and vice versa. Moreover, Here are a couple of use cases for locally scoped class and globally scoped class.

- **switch on** global scope for this selector inside the local selector. For example, in order to take full advantage of a globally existed class (as mentioned following) and avoid bloated css file, you need to apply a global class to a component, that is in local scope.

```css
.element {
  & :global .global-class-name {}
}
```

- **switch on** local scope for this selector inside the global selector.

```css
.element {
  & :local .local-class-name {}
}
```
Last but not least, the syntax `:global (without brackets)` is used to **switch on** global mode for the selector above. As mentioned above, it is suitable for some scenario like a mode switch. In addition, and if you are attentive, you can find the final class names in an outputted css file are defined with the value of `:global(.global-class-name)` or `:local(.local-class-name)`.

**Comment by liuliangsir** [comment]
That's absolutely right, all classes are local by default. What if you would like to apply global scope to this selector inside a local selector? In this case,  you can use the syntax `:global (without brackets)` for this selector above and vice versa. Moreover, Here are a couple of use cases for locally scoped class and globally scoped class.

- **switch on** global scope for this selector a the local selector. For example, in order to take full advantage of a globally existed class (as mentioned following) and avoid bloated css file, you need to apply a global class to a component, that is in local scope.

```css
.element {
  & :global .global-class-name {}
}
```

- **switch on** local scope for this selector inside the global selector.

```css
.element {
  & :local .local-class-name {}
}
```
Last but not least, the syntax `:global (without brackets)` is used to **switch on** global mode for the selector above. As mentioned above, it is suitable for some scenario like a mode switch. In addition, and if you are attentive, you can find the final class names in an outputted css file are defined with the value of `:global(.global-class-name)` or `:local(.local-class-name)`.

**Comment by gaurav5430** [comment]
If **modules: 'global'**, in webpack css-loaders config, all classes are global by default, so if you have to locally scope some classes, you have to use **":local"**
similarly if **modules: 'true' //local by default**, you have to explicitly use **":global"** to globally scope some classes.

There is still some confusion for me between **:global(.className)** vs **:global .className**, the first one seems to be used for declaring a global class,
but does the second one only temporarily change the scope? can we later change the scope back to local ?

**Comment by juliofuentesUI** [comment]
> If **modules: 'global'**, in webpack css-loaders config, all classes are global by default, so if you have to locally scope some classes, you have to use **":local"**
> similarly if **modules: 'true' //local by default**, you have to explicitly use **":global"** to globally scope some classes.
> 
> There is still some confusion for me between **:global(.className)** vs **:global .className**, the first one seems to be used for declaring a global class,
> but does the second one only temporarily change the scope? can we later change the scope back to local ?

Thank you for the beautiful simplified explanation! I don't understand why they have to over complicate what should be a straight forward concept.

**Comment by gaurav5430** [comment]
Found some more info regarding `:global .className` vs `:global(.className)`

- `:global(.className)` would declare a global class. 
When do you need to declare a global class? 
Well, I needed to do it when I wanted my classes to not be hashed, so that I can use them in regular css `className` instead of the `styles.className` syntax, but I still needed to declare them in css-modules way using `.global(.className)`, so that I can still use them with css modules to compose local classes from those global classes.

- `:global .className` or `.className :global` would mean that any classnames or declarations you refer inside this block would be used from global declarations.
For this, the only two examples i have found/figured out:
1. animations - when you declare animations globally, if you need to refer them from some local class you have to use .className :global { animation-name: global-animation-name )

- One way is to declare the animation in the same file and use it as you usually do.

**declaration:**
`@keyframes animation-1 { ... }`

**usage:**
```
.className {
    animation-name: animation-1 
}
```

- or you can declare your animations globally

**declaration:**
`@keyframes :global(animation-1) { ... }`

**usage:**
```
.className :global {
    animation-name: animation-1
}
```

2. targeting a regular global class from within a local class.
`.c1 :global .c2 { ... }`
here c1 is compiled through css modules and is locally scoped, but c2 is a regular css class, which is global by default.
For more info about this usage, see the Global Selectors section here: https://css-modules.github.io/webpack-demo/

**Comment by nairsirk** [comment]
@gaurav5430 sorry, i still don't get it on declaring `.global(.className)` for composing local classes from global classes. what `.global` used for? can you tell more about it?

**Comment by dfdeagle47** [comment]
> @gaurav5430 sorry, i still don't get it on declaring `.global(.className)` for composing local classes from global classes. what `.global` used for? can you tell me more about it?

One contrived example I can think of is if

- if you have `.class1` that is a class that's declared in a third-party lib you're importing, so you might have no control over how it's declared and it might be a global class.
- and you have `.class2` which is a local class applied to one of your components

So you might need to have the rules

```css
:global(.class1) .class2 {
  /* ...CSS rules... */
}
```

A similar example I had in the past is


```css
.modalWrapper {
  :global(.modal) {
    max-width: 600px;
  }
}
```

The `.modalWrapper` class is a local class I had created and the global `.modal` was a class which came from a third-party package.


Does that make sense?

Though I think it's not something you will use regularly, and it's just good to know it's there if you encounter an issue with global vs local classes.

**Comment by gaurav5430** [comment]
> @gaurav5430 sorry, i still don't get it on declaring `.global(.className)` for composing local classes from global classes. what `.global` used for? can you tell me more about it?

This is my understanding:
if you are using css modules in the 'local by default' mode, it will hash all the classes once you import the css file to javascript. But if you want some class to not be hashed, you would need to declare it as global by using the :global() syntax. This would mean that css modules would not hash these classes, and you are free to use them as regular css, without needing to import them in each js file where you want to use them, through css modules. You can just import it once at the top level js file and now they would be available to use everywhere in your app as regular global css.
I have some global classes in my app, for e.g `m-16` for margin 16px, now when I declare local classes in my component css, sometimes i compose them from these global classes, say if a component requires margin 16px, instead of declaring it locally I would compose the global `m-16` class in the local class. I don't have to compose from a particular filename here because css modules now knows that these are global classes, so i can just say `composes: m-16 from global`

**Comment by nairsirk** [comment]
@dfdeagle47 @gaurav5430 okay , i got it . thank you guys , really appreciate it

**Comment by Hedes-Productions** [comment]
@dfdeagle47  I got a problem like this, I had a Bootstrap component named carousel-control-next. So this cannot be accessed normally because it has no control over how it's declared. So I needed to give different two types of stylings by making two different CSS module files. So as first, I used `:global` . But then realized that it's going to affect all the files. But I needed to control that class component differently in those CSS files. Is there any way to do that?

**Comment by dfdeagle47** [comment]
@Hedes-Productions 

> @dfdeagle47 I got a problem like this, I had a Bootstrap component named carousel-control-next. So this cannot be accessed normally because it has no control over how it's declared. So I needed to give different two types of stylings by making two different CSS module files. So as first, I used `: global`. But then realized that it was going to affect all the files. But I needed to control that class component differently in those CSS files(that class cannot be controlled normally). Is there any way to do that?

Since I don't have the code, it's hard to give you a correct answer for sure, but idea would be to do something like I described in [my comment above](https://github.com/css-modules/css-modules/issues/264#issuecomment-812095155): to add another class to your HTML carrousel component, and use that class to override the default styling.

So something like:

```css
.styling-1 :global(.carousel-control-next) {
  /* ...CSS rules for styling 1... */
}

.styling-2 :global(.carousel-control-next) {
  /* ...CSS rules for styling 2... */
}
```

and you would inject the local classes `.styling-1` or `.styling-2` onto you carrousel HTML component which contains the global class ".carousel-control-next" as one of its children.

**Comment by rmirro** [comment]
I've just done some experimenting with this and the difference between: 
- default `local` scope 
- `:global(<selector>)`
- `:global <selector>` 
 
seems to be how deeply `:global` is applied.

---

### file: `test-component.module.scss`

- ### default `local` scope:
```css
.component {
  padding: 1rem;

  .nested {
    padding: 2rem;

    .nested-nested {
      padding: 3rem;
    }
  }
}


.test-component-module_component__NyYUN {
  padding: 1rem;
}
.test-component-module_component__NyYUN .test-component-module_nested__NU9QJ {
  padding: 2rem;
}
.test-component-module_component__NyYUN .test-component-module_nested__NU9QJ .test-component-module_nested-nested__94M-i {
  padding: 3rem;
}
```

- ### `:global(<selector>)`
```css
.component {
  padding: 1rem;

  :global(.nested) {
    padding: 2rem;

    .nested-nested {
      padding: 3rem;
    }
  }
}


.test-component-module_component__NyYUN {
  padding: 1rem;
}
.test-component-module_component__NyYUN .nested {
  padding: 2rem;
}
.test-component-module_component__NyYUN .nested .test-component-module_nested-nested__94M-i {
  padding: 3rem;
}
```

---

- ### `:global <selector>`
```css
.component {
  padding: 1rem;

  :global .nested {
    padding: 2rem;

    .nested-nested {
      padding: 3rem;
    }
  }
}

.test-component-module_component__NyYUN {
  padding: 1rem;
}
.test-component-module_component__NyYUN .nested {
  padding: 2rem;
}
.test-component-module_component__NyYUN .nested .nested-nested {
  padding: 3rem;
}
```
I'm leaning towards going with `:global <selector>` for all react components because it seems to allow the benefits of encapsulation via CSS Modules while minimizing clumsy-ish `classNames` everywhere:
```js
import styles from './test-component.module.scss';

export const TestComponent = () => {
  return (
    <div className={styles['component']}>
      <div className="nested">
        <div className="nested-nested"></div>
      </div>
    </div>
  );
}
```

**Comment by grumd** [comment]
> I'm leaning towards going with :global <selector> for all react components

I'd prefer not to do that. `:global` should be used only for 3rd party classes or global selectors, for two reasons, 1st being that you can use Typescript to avoid typos in classnames if all classes you use come from css-modules, and 2nd being that you can avoid potentially colliding with an actually globally scoped selector, e.g. if you have both `.local__bUao5 > .button` and global `.button` somewhere.

If you use camelCase, your classes aren't that ugly in jsx anyway. `<div className={styles.component}><div className={styles.nestedNested}></div></div>`

**Comment by rmirro** [comment]
I'm not seeing any TS support in vscode for class names that seems like it could assist in typo avoidance...
```js
import styles from './my-component.module.scss';
```
Appears to be typed as an index signature:
```js
declare module '*.module.scss' {
  const classes: { readonly [key: string]: string };
  export default classes;
}
```
I prefer not to use camelCase class names but agree that it could make references less awkward if it's supported (it doesn't seem that it is with the out-of-the-box nx/react configuration I'm using).

If using ultra generic classes like `.button`, then yes, that seems potentially problematic. However, simplistic name spacing seems like it goes a long way towards avoiding that sort of issue without having to locally scope every nested class.

Whatever you're comfortable with I guess...  😄

**Comment by grumd** [comment]
@rmirro TS support is done with a plugin: https://github.com/mrmckeb/typescript-plugin-css-modules

When working in a big team I'd rather use css-modules or other tools, than hope every single person on the team never makes a classname that's too generic.

I'm not gonna bash your preference in how to use css-modules, just sharing my point of view though.

**Comment by rmirro** [comment]
Team size could become nearly irrelevant depending on the effectiveness of linting and the code review process.  Luckily _hope_ hasn't been a factor of development with the past few teams I've worked with.

We're re-rewriting a stencil app in react and css-modules is helping provide automated encapsulation similar-ish to what can be expected from web components/shadow dom.

Each primary component stylesheet references a sass mixin that allows that component to namespace itself via css-modules but also have internal _global_ (to this component) classes.  This is meant as a convenience factor, as well as to promote styling overrides of shared components.

We have since adopted MUI, which seems to have taken a similar approach:
```html
<button class="MuiButton-root MuiButton-contained css-1hw9j7s">
```

Our components are only consumed by other applications internal to our company, so although there is no absolute guarantee that one of their global classes might be applied to our component, it's highly unlikely thanks the simplistic namespacing (similar to `MuiButton-` above) that is also in place.

**Comment by HasanMothaffar** [comment]
> I've just done some experimenting with this and the difference between:
> 
>     * default `local` scope
> 
>     * `:global(<selector>)`
> 
>     * `:global <selector>`
> 
> 
> seems to be how deeply `:global` is applied.
> ### file: `test-component.module.scss`
> 
>     * ### default `local` scope:
> 
> 
> ```css
> .component {
>   padding: 1rem;
> 
>   .nested {
>     padding: 2rem;
> 
>     .nested-nested {
>       padding: 3rem;
>     }
>   }
> }
> 
> 
> .test-component-module_component__NyYUN {
>   padding: 1rem;
> }
> .test-component-module_component__NyYUN .test-component-module_nested__NU9QJ {
>   padding: 2rem;
> }
> .test-component-module_component__NyYUN .test-component-module_nested__NU9QJ .test-component-module_nested-nested__94M-i {
>   padding: 3rem;
> }
> ```
> 
>     * ### `:global(<selector>)`
> 
> 
> ```css
> .component {
>   padding: 1rem;
> 
>   :global(.nested) {
>     padding: 2rem;
> 
>     .nested-nested {
>       padding: 3rem;
>     }
>   }
> }
> 
> 
> .test-component-module_component__NyYUN {
>   padding: 1rem;
> }
> .test-component-module_component__NyYUN .nested {
>   padding: 2rem;
> }
> .test-component-module_component__NyYUN .nested .test-component-module_nested-nested__94M-i {
>   padding: 3rem;
> }
> ```
> 
>     * ### `:global <selector>`
> 
> 
> ```css
> .component {
>   padding: 1rem;
> 
>   :global .nested {
>     padding: 2rem;
> 
>     .nested-nested {
>       padding: 3rem;
>     }
>   }
> }
> 
> .test-component-module_component__NyYUN {
>   padding: 1rem;
> }
> .test-component-module_component__NyYUN .nested {
>   padding: 2rem;
> }
> .test-component-module_component__NyYUN .nested .nested-nested {
>   padding: 3rem;
> }
> ```
> 
> I'm leaning towards going with `:global <selector>` for all react components because it seems to allow the benefits of encapsulation via CSS Modules while minimizing clumsy-ish `classNames` everywhere:
> 
> ```js
> import styles from './test-component.module.scss';
> 
> export const TestComponent = () => {
>   return (
>     <div className={styles['component']}>
>       <div className="nested">
>         <div className="nested-nested"></div>
>       </div>
>     </div>
>   );
> }
> ```

Thank you! This makes total sense to me. Your answer should be featured in the docs in my opinion.. I've been struggling to find a clear answer

**Comment by Andrey-Pavlov** [comment]
@rmirro I spent several hours to understand how global and local classes actually work until I've found your answer. Thank you.
